src/hotspot/share/logging/logOutputList.cpp

```diff
@@ -127,6 +127,10 @@ void LogOutputList::add_output(LogOutput* output, LogLevelType level) {
        node->_next = node->_next->_next) {
   }
 
+  // Prevent the reordering of adding node to list and setting the value of node.
+  // Such a reordering would lead to reading incorrect values.
+  OrderAccess::release();
```

Comment:
I agree this is the spot-on. what we need is a storestore barrier.  

We don't want to 'node->_value = output;' floats below stores, such as '_level_start[l] = node;'

I meant to ask whether 'storestore' barrier itself can resolve this intermittent error on your side?  I realized that your had storestore barrier in the early revision. 

>  By the way, I have tested millions of times and it hasn't failed when just adding memory barrier on the writing side.

So you have verified it, haven't you?

I can help you fill in some backgrounds about why it's not mandatory to use release-acquire.  this RCU hack starts from Linux kernel...

Reply:
@navyxliu  thank you for looking at this. As you note the RCU code is intended to ensure that a reader can either see the old node, or the new node, and work correctly. In that case  we would logically expect to use release semantics so that if a reader sees the new node then it also sees the correct values for all fields of the node. The release barrier is a `storeStore|loadStore` barrier. Release semantics then require matching acquire semantics when we read the node from the list. You seem to be suggesting that RCU obviates the need for the acquire side of this - is that correct? If so then perhaps the raw storeStore barrier will suffice - as we state in orderAccess.hpp:
> If not used in such a pair, it is advised to use a membar instead: acquire/release only make sense as pairs.

Personally I would find the acquire/release easier to reason about as I'm not that familiar with the subtleties of RCU. That said I'm also concerned that this RCU implementation is again found to be flawed due to missing barriers and have to wonder if this is the last of them? If RCU is too difficult to get right then perhaps we should stop trying to be overly clever and use a simpler, potentially slower, lock-based solution?

Reply:
> > By the way, I have tested millions of times and it hasn't failed when just adding memory barrier on the writing side.
> 
> So you have verified it, haven't you?

:-)
 
> I can help you fill in some backgrounds about why it's not mandatory to use release-acquire. this RCU hack starts from Linux kernel...
> 
> It doesn't use explicit synchronization for scalability. 

Kernel RCU depends on assumptions not guaranteed by any language standard.  In particular, it depends on read dependency ordering. This works on most processors unless the compiler does something you weren't expecting it to do. But, of course, it is undefined behaviour according to the language. The Linux kernel itself uses some compiler-specific black magic to do the needful: see `READ_ONCE` in include/linux/compiler.h. 

We might decide to do something similar for really hot code, but this isn't it.

Reply:
@dholmes-ora 
> You seem to be suggesting that RCU obviates the need for the acquire side of this - is that correct? If so then perhaps the raw storeStore barrier will suffice - as we state in orderAccess.hpp

yes, that's my understanding.  Release has compound effect, but I still think we only need a storestore barrier.  In the runtime, there's only one active writer.  Other writers are all blocked by ConfigurationLock.  

Even in weak memory model, we don't need to worry about loadStore order because they are either control dependency (line 8) or true data dependency (line 9). 

```
  1 void LogOutputList::add_output(LogOutput* output, LogLevelType level) {
  2   LogOutputNode* node = new LogOutputNode();
  3   node->_value = output;
  4   node->_level = level;
  5
  6   // Set the next pointer to the first node of a lower level
  7   for (node->_next = _level_start[level];
  8        node->_next != nullptr && node->_next->_level == level;
  9        node->_next = node->_next->_next) {
 10   }
 11 ...
 12   a storestore barrier
 13   publish 'node'
```

however, we need a storestore barrier at line 12 on all platforms indeed. Even though x86_64 will never reorder 2 stores, C/C++ compiler could. 

I tried 5000 times on a 64-cores arm64 instance, but I can't trigger this error. 
@gaogao-mem what's your preference?  If a storestore barrier can resolve your error, do we want to go to release/acquire pair? 

>  If RCU is too difficult to get right then perhaps we should stop trying to be overly clever and use a simpler, potentially slower, lock-based solution?

The current solution is there before I added 2 tests in JDK-8267952. The unit tests just simulate 'log dmcd' in real life. I look up JBS, 8305819 is the first record we encounter this error. maybe the new microarchitecture of arm64 reorders memory operations more aggressively than before?   

if we replace the current solution with read/write lock, I guess we won't lose much in terms of performance. but this seems a separated task.

Reply:
@theRealAph I didn't mean that current code is like linux kernel RCU, which is well-documented synchronization. I mean current solution in UnifiedLog borrow the idea. 

LogOutputList::increase_readers() is similar to rcu_read_lock(). 

I found READ_ONCE in 'include/asm-generic/rwonce.h'. In hotspot, we could use it for case-2: 
```
(2) Ensuring that the compiler does not fold, spindle, or otherwise
 * mutilate accesses that either do not require ordering or that interact
 * with an explicit memory barrier or atomic instruction that provides the
 * required ordering.
```

Analysis so far leads us to believe this error manifests in weak memory model, right?  it's the processor reorders some instructions but not compiler optimizations. Do we really need to consider that MACRO?

Reply:
> @theRealAph I didn't mean that current code is like linux kernel RCU, which is well-documented synchronization. I mean current solution in UnifiedLog borrow the idea.

Yes, it does borrow, the idea, but AFAICS without understanding just how tricky it is. And gets it all wrong as a consequence.

> Analysis so far leads us to believe this error manifests in weak memory model, right? it's the processor reorders some instructions but not compiler optimizations. Do we really need to consider that MACRO?

I don't think that's the real problem. While HotSpot will very probably never be standard C++, we usually try to write standard-correct code, going outside the standard guarantees when we really need to.  I don't think this is such a case.

Data races are undefined behaviour. With undefined behaviour, all bets are off. You cannot test for the absence of bugs, only their presence.

Reply:
> if we replace the current solution with read/write lock, I guess we won't lose much in terms of performance. but this seems a separated task.

Let's do that. Then we'll have code that's known to be correct, and we can all do other things. 😁

Reply:
> > if we replace the current solution with read/write lock, I guess we won't lose much in terms of performance. but this seems a separated task.
> 
> Let's do that. Then we'll have code that's known to be correct, and we can all do other things. grin

https://github.com/openjdk/jdk/pull/9838

Here's my old PR for a read-write lock that was meant to replace this RCU mechanism :-). It's a simple counter based one, which I borrowed from ZGC. If anyone wants to pick up this code then feel free to do so (with me as contributor and Oracle copyright notice).

Reply:
@jdksjolen Could you please integrate #9838 to master and then we can apply the read-write lock in the UL?

Reply:
The PR was never approved and was somewhat contentious. I would want to see how UL uses the proposed RWL before considering the integration of both changes. This is not a trivial amount of work I know.

Buried on the other PR was a suggestion to switch the broken UL RCU mechanism for one of the other VM RCU-like mechanisms, but I'm not familiar enough with those other mechanisms to say if that is viable. But perhaps the UL RCU mechanism can be examined against those other ones to see if any more bugs are evident?

Or we just proceed with a suitably commented storeStore barrier and hope that this is the last RCU bug.

Reply:
> Or we just proceed with a suitably commented storeStore barrier and hope that this is the last RCU bug.

We can't do only that, because it's still buggy. For it to be correct portable code we need at least a `memory_order::consume` load on the reading side. Given that `consume` is C++20 and we're still on C++14, something like the kernel's `READ_ONCE` would probably do.

Reply:
I assume that is the memory barrier that RCU is supposed to be implicitly implementing? If this is missing from this RCU-like thing can we just use release/acquire as suggested much earlier?

Reply:
hi, @theRealAph 

I learn your comments and Boehm's. Actually, I found that a discussion on control dependency ordering took place in the PR of JDK-8288904 before. The knowledge is non-trivial but I think I got your point.

IMHO, you insist the general rule to a niche case. you are still right, but now the bar is too high to resolve this immediate error. 

1. compiler may flip orders regardless control dependency -- check! you also makes reference of kernel RCU. However, UL configuration is much simpler than C++ standard libraries and Linux kernel. In this case, I think we should take the advice of "Recipient only reads x" from Boehm.  there's a single writer. it's manipulating linked-list. within a CPU, control and data dependency is guaranteed. 

2. I believe that the error reported in JDK-8305819 is caused by hardware instead of compiler optimization.  Evidence: I was told by @gaogao-mem this intermittent crash disappeared after he integrated a storestore barrier. He has ran the tests millions of times and the error never occurs again.  Can we settle that a store-only barrier can practically resolve this error? 

We also provided reasoning why we need a storestore barrier in prior comments. we will put comments in PR. The process is same as JDK-8288904. Reason, merge and resolve the error practically. 

3. I am not saying it's perfect. I agree the current solution is still "buggy" in presence of compiler optimizations.  we never enforce it to the C++ standard so it's "UB". 

That's why we should work on replacing RCU with RWL. Like David pointed out, we need more time to review, test a solution. To achieve that, we need a better baseline. now we know that the concurrent test is broken on the weak memory model. If this bandaid works, I think we will be in the better position for the development.

Reply:
> I assume that is the memory barrier that RCU is supposed to be implicitly implementing? If this is missing from this RCU-like thing can we just use release/acquire as suggested much earlier?

currently, it uses Atomic::add in LogOutputList::increase_readers/descrease_readers. The default order of Atomic::add is `memory_order_conservative` so it emits a full memory barrier.

Reply:
After @theRealAph 's detailed explanation and reading some related materials, I am convinced that we should use acquire operation on the read side, and I'd like to explain why in detail based on my current understanding.

I suppose that we have agreed that a storestore barrier is needed on the write side.

We can simplify the logic of the problematic region as follows:

```

                   The Writer                        |      The Reader(s)
                                                     |
(1) create a new node                                |  (1) traverse all nodes from _level_start[level]
(2) init the node (set _value, _level, _next ...)    |  (2) call (node->_value)->write(...)
(3) update the _level_start array                    |
(4) link the node to the list                        |
```

The writer step (3) and (4) add the new node. The reader step (1) may get the new node(s) added by the writer and then call (node->value)->write(...).

According to the [std::memory_order page](https://en.cppreference.com/w/cpp/atomic/memory_order) :

```
If one evaluation modifies a memory location, and the other reads or modifies the same memory location,
and if at least one of the evaluations is not an atomic operation, the behavior of the program is
undefined (the program has a data race) unless there exists a happens-before relationship between
these two evaluations.
```

There is a data race between the writer and reader(s). The writer step (3) and (4) modify the _level_start array and _next of a old node, and the reader step (1) reads from _level_start and _next of a node. This causes a UB. And in current problem, the reader step (2) may read an incorrect _value from a new node.

To fix this problem, we need to establish a happens-before relationship (Inter-thread happens-before). That means when step (1) on read side get a new node, we must make sure that step (1) and (2) on the write side happens-before step (2) on read side.

We can achieve the goal by `Dependency-ordered before` since the reader step (1) carries a dependency into step (2). Unfortunately, Hotspot currently doesn't support consume operation just as Aph metioned, so we should depend on `Synchronizes with`.

Combining the [std::memory_order page](https://en.cppreference.com/w/cpp/atomic/memory_order) and [std::atomic_thread_fence page](https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) together, we can get a solution: add a release fence between step (2) and (3) on write side and use acquire operation on read side step (1).

At first, I only added a storestore barrier on the write side. The reason is reader step (1) and (2) have a dependency and the only CPU I know that may reorder dependent loads is Alpha, and Alpha has gone.

Because the code here is shared and we cannot assume the architecture we support, the acquire operation is necessary according to the language standard and the current implementation of hotspot, and also we need to make _level_start and _next volatile.

The above is my current understanding. Please feel free to correct me if I'm wrong.

This is the only place I can find now that has data race. Should we consider integrating this patch first if the current solution is acceptable and then rewrite the logic by RWL after RWL is supported by hotspot?

Reply:
Hi,

> I learn your comments and Boehm's. Actually, I found that a discussion on control dependency ordering took place in the PR of JDK-8288904 before. The knowledge is non-trivial but I think I got your point.

Well, maybe.

I think the core problem here is that the C++ memory model is clear enough, but many people want to find ways to cheat it. "Hey, can't we rely on a data dependency?" The answer to; that question is certainly no, you can't, not until we use `memory_order_consume`.

>     within a CPU, control and data dependency is guaranteed.

No, it really isn't. This is shared code, not CPU-specific, and therefore we don't know anything about any CPU.
 
>     2. I believe that the error reported in JDK-8305819 is caused by hardware instead of compiler optimization.  

In this case, yes, but the real error is between keyboard and chair: the programmer.

I believe you're thinking about this problem at entirely the wrong level. This is a C++ program, and the compiler interprets it according to the rules of C++. Therefore, the only rules to which anyone may appeal are the rules of C++.

> Evidence: I was told by @gaogao-mem this intermittent crash disappeared after he integrated a storestore barrier. He has ran the tests millions of times and the error never occurs again.  Can we settle that a store-only barrier can practically resolve this error?

I don't understand your reasoning. Why do you want to cheat? Why not simply follow the rules?

> We also provided reasoning why we need a storestore barrier in prior comments. we will put comments in PR. The process is same as JDK-8288904. Reason, merge and resolve the error practically.

>     3. I am not saying it's perfect. I agree the current solution is still "buggy" in presence of compiler optimizations.  we never enforce it to the C++ standard so it's "UB".

So, if you know that it's still incorrect, what is your motive for pushing an incorrect change, when the correct one is simple? Why not push a correct change today instead? We know what it is, let's simply use proper synchronization and no-one will ever have to look at this again, regardless of compiler and CPU changes.

Reply:
> The PR was never approved and was somewhat contentious. I would want to see how UL uses the proposed RWL before considering the integration of both changes. This is not a trivial amount of work I know.
> 
> Buried on the other PR was a suggestion to switch the broken UL RCU mechanism for one of the other VM RCU-like mechanisms, but I'm not familiar enough with those other mechanisms to say if that is viable. But perhaps the UL RCU mechanism can be examined against those other ones to see if any more bugs are evident?
> 
> Or we just proceed with a suitably commented storeStore barrier and hope that this is the last RCU bug.

Yes, the VM has a (correct, AFAIK) RCU mechanism of its own in `GlobalCounter`. As the name implies, it's a global construct that's shared by all users of the mechanism. If a lot of logging occurs, then UL might put a strain on the mechanism that we haven't seen before and this in turn might have performance implications.